src/plan/generational/immix/global.rs

```diff
@@ -0,0 +1,288 @@
+                .schedule_common::<GenImmixMatureProcessEdges<VM, { TraceKind::Defrag }>>(
+                    &GENIMMIX_CONSTRAINTS,
+                    scheduler,
+                );
+            // Stop & scan mutators (mutator scanning can happen before STW)
+            scheduler.work_buckets[WorkBucketStage::Unconstrained].add(StopMutators::<
+                GenImmixMatureProcessEdges<VM, { TraceKind::Defrag }>,
+            >::new());
+        } else {
+            debug!("Full heap GC Fast");
+            self.common()
+                .schedule_common::<GenImmixMatureProcessEdges<VM, { TraceKind::Fast }>>(
+                    &GENIMMIX_CONSTRAINTS,
+                    scheduler,
+                );
+            // Stop & scan mutators (mutator scanning can happen before STW)
+            scheduler.work_buckets[WorkBucketStage::Unconstrained].add(StopMutators::<
+                GenImmixMatureProcessEdges<VM, { TraceKind::Fast }>,
+            >::new());
+        }
+
+        // Prepare global/collectors/mutators
+        scheduler.work_buckets[WorkBucketStage::Prepare]
+            .add(Prepare::<Self, GenImmixCopyContext<VM>>::new(self));
+        if is_full_heap {
+            if defrag {
+                scheduler.work_buckets[WorkBucketStage::RefClosure].add(ProcessWeakRefs::<
+                    GenImmixMatureProcessEdges<VM, { TraceKind::Defrag }>,
+                >::new());
+            } else {
+                scheduler.work_buckets[WorkBucketStage::RefClosure].add(ProcessWeakRefs::<
+                    GenImmixMatureProcessEdges<VM, { TraceKind::Fast }>,
+                >::new());
+            }
+        } else {
+            scheduler.work_buckets[WorkBucketStage::RefClosure].add(ProcessWeakRefs::<
+                GenNurseryProcessEdges<VM, GenImmixCopyContext<VM>>,
+            >::new());
+        }
+        // Release global/collectors/mutators
+        scheduler.work_buckets[WorkBucketStage::Release]
+            .add(Release::<Self, GenImmixCopyContext<VM>>::new(self));
+        // Resume mutators
+        #[cfg(feature = "sanity")]
+        {
+            use crate::util::sanity::sanity_checker::*;
+            scheduler.work_buckets[WorkBucketStage::Final]
+                .add(ScheduleSanityGC::<Self, GenImmixCopyContext<VM>>::new(self));
+        }
+        scheduler.set_finalizer(Some(EndOfGC));
+    }
+
+    fn get_allocator_mapping(&self) -> &'static EnumMap<AllocationSemantics, AllocatorSelector> {
+        &*super::mutator::ALLOCATOR_MAPPING
+    }
+
+    fn prepare(&mut self, tls: VMWorkerThread) {
+        let full_heap = !self.is_current_gc_nursery();
+        self.gen.prepare(tls);
+        if full_heap {
+            self.immix.prepare(full_heap);
+        }
+    }
+
+    fn release(&mut self, tls: VMWorkerThread) {
+        let full_heap = !self.is_current_gc_nursery();
+        self.gen.release(tls);
+        if full_heap {
+            let did_defrag = self.immix.release(full_heap);
```

Comment:
Yes, it is possible. We can call `self.immix.in_defrag()` before calling `release()`. Otherwise, the defrag value will be set to `false` in `release()`. 

I was following the Java MMTk code and added the return value. I feel keeping the same patterns as in the Java MMTk can help us debugging (we can easily refer to the Java MMTk code), and can help us when we port more code from Java MMTk.

If we would like to have a trait method for `release`, such as `Space.release(&self, full_heap: bool)`, we will have to remove the return value. Is there any particular reason that you suggested removing the return value?

Reply:
I see. Then it's better to follow the Java code.